Return of a function used as an argument generates C compilation error.

<!-- Please make sure to run `v up` before reporting any issues as it may have already been fixed.
     It's also advisable to update all relevant modules using `v outdated` and `v install` -->

<!-- You can use `v doctor` to fill up the next fields -->
**V version:** `0.2.4 fd34ebd.6137ce2`
**OS:** `linux, Linux version 5.16.8-arch1-1 (linux@archlinux) (gcc (GCC) 11.1.0, GNU ld (GNU Binutils) 2.36.1) #1 SMP PREEMPT Tue, 08 Feb 2022 21:21:08 +0000 (VM)`

<!-- Please include the standalone minimal reproducible code.
If not, it may be closed by the administrators. -->
**What did you do?**
```v
    todo_dto := json.decode(TodoDto, json.encode(body)) or {
      res.send('Bad Request', 400)
      return
    }
```
 
**What did you see instead?**
```
/tmp/v_1000/v-todo.7447070227465168314.tmp.c:11355: error: too few arguments to function
builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang
```

## But I found solution!
If I will write
```v
    body_jsoned := json.encode(body)
    todo_dto := json.decode(TodoDto, body_jsoned) or {
      res.send('Bad Request', 400)
      return
    }
```
then it compiles without error. I don't know if it matters but I do it in a closure.  
Here is full code of it [https://github.com/ScuroGuardiano/v-todo/blob/3f39752d73f7414eb806e398527eddde97b7f2b1/routers/todo.v#L37](https://github.com/ScuroGuardiano/v-todo/blob/3f39752d73f7414eb806e398527eddde97b7f2b1/routers/todo.v#L37)

Minimal test case:
```v
import json

struct TodoDto {}
body := ''
_ := json.decode(TodoDto, json.encode(body)) or {
    return
}
```
C error:
```
$ ./v -cc gcc ../old/jsondecode.v
==================
C:\Users\CZASTK~1\AppData\Local\Temp\v_0\jsondecode.14855224073957659675.tmp.c: In function 'main__main':
C:\Users\CZASTK~1\AppData\Local\Temp\v_0\jsondecode.14855224073957659675.tmp.c:11809:15: error: too few arguments to function 'json__encode_string'
  cJSON* _t3 = json__encode_string();
               ^~~~~~~~~~~~~~~~~~~
C:\Users\CZASTK~1\AppData\Local\Temp\v_0\jsondecode.14855224073957659675.tmp.c:11780:31: note: declared here
 VV_LOCAL_SYMBOL struct cJSON* json__encode_string(string val) {
                               ^~~~~~~~~~~~~~~~~~~
C:\Users\CZASTK~1\AppData\Local\Temp\v_0\jsondecode.14855224073957659675.tmp.c:11816:29: error: invalid initializer
  Option_main__TodoDto _t1 = (102 /* main.TodoDto */, (_t7));
                             ^
```

--- a/vlib/json/json_decode_with_encode_arg_test.v
+++ b/vlib/json/json_decode_with_encode_arg_test.v
@@ -1,0 +1,12 @@
+import json
+
+struct TodoDto {
+	foo int
+}
+
+fn test_decode_with_encode_arg() ? {
+	body := TodoDto{}
+	ret := json.decode(TodoDto, json.encode(body)) ?
+	println(ret)
+	assert ret.foo == 0
+}

--- a/vlib/json/json_decode_with_optional_arg_test.v
+++ b/vlib/json/json_decode_with_optional_arg_test.v
@@ -1,0 +1,19 @@
+import json
+import os
+
+struct DbConfig {}
+
+fn test_json_decode_with_optional_arg() {
+	if ret := print_info() {
+		println(ret)
+	} else {
+		println(err)
+	}
+	assert true
+}
+
+fn print_info() ?string {
+	dbconf := json.decode(DbConfig, os.read_file('dbconf.json') ?) ?
+	println(dbconf)
+	return '$dbconf'
+}

--- a/vlib/json/json_decode_with_optional_arg_test.v
+++ b/vlib/json/json_decode_with_optional_arg_test.v
@@ -1,7 +1,9 @@
 import json
 import os
 
-struct DbConfig {}
+struct DbConfig {
+	foo int
+}
 
 fn test_json_decode_with_optional_arg() {
 	if ret := print_info() {

--- a/vlib/v/gen/c/fn.v
+++ b/vlib/v/gen/c/fn.v
@@ -1180,10 +1180,11 @@ fn (mut g Gen) fn_call(node ast.CallExpr) {
 	is_json_encode := name == 'json.encode'
 	is_json_encode_pretty := name == 'json.encode_pretty'
 	is_json_decode := name == 'json.decode'
-	g.is_json_fn = is_json_encode || is_json_encode_pretty || is_json_decode
+	is_json_fn := is_json_encode || is_json_encode_pretty || is_json_decode
 	mut json_type_str := ''
 	mut json_obj := ''
-	if g.is_json_fn {
+	if is_json_fn {
+		g.is_json_fn = true
 		json_obj = g.new_tmp_var()
 		mut tmp2 := ''
 		cur_line := g.go_before_stmt(0)
@@ -1218,7 +1219,6 @@ fn (mut g Gen) fn_call(node ast.CallExpr) {
 			// its name was already used to generate the function call
 			g.is_js_call = true
 			g.call_args(node)
-			g.is_js_call = false
 			g.writeln(');')
 			tmp2 = g.new_tmp_var()
 			g.writeln('Option_$typ $tmp2 = $fn_name ($json_obj);')
@@ -1383,7 +1383,7 @@ fn (mut g Gen) fn_call(node ast.CallExpr) {
 			}
 			mut tmp_cnt_save := -1
 			g.write('(')
-			if g.is_json_fn {
+			if is_json_fn {
 				g.write(json_obj)
 			} else {
 				if node.is_keep_alive
@@ -1546,6 +1546,7 @@ fn (mut g Gen) call_args(node ast.CallExpr) {
 		if node.args.len < 1 {
 			g.error('node should have at least 1 arg', node.pos)
 		}
+		g.is_js_call = false
 		node.args[1..]
 	} else {
 		node.args

